### PR TITLE
day2

### PR DESCRIPTION
**Pull request by TaehoSon-hds**

### CODE REVIEW

**Comment by TaehoSon-hds** [comment]
```ssh
10-days-of-grad on  day2 [!] via λ 9.2.8 via ❄️  impure (ghc-shell-for-ten-days-of-grad-0.5-0-env) took 6s 
✦ ❯ nix build
warning: Git tree '/home/sth/Projects/10-days-of-grad' is dirty

10-days-of-grad on  day2 [+] via λ 9.2.8 via ❄️  impure (ghc-shell-for-ten-days-of-grad-0.5-0-env) took 18s 
✦ ❯ ./result/bin/nn 
Circles problem, 1 hidden layer of 128 neurons, 1000 epochs
---
Training accuracy (gradient descent) 93.0
Validation accuracy (gradient descent) 95.0

Training accuracy (Adam) 99.0
Validation accuracy (Adam) 98.0


Spirals problem, Adam, 700 epochs
---
1 hidden layer, 128 neurons (513 parameters)
Training accuracy 82.2
Validation accuracy 72.0

1 hidden layer, 512 neurons (2049 parameters)
Training accuracy 88.8
Validation accuracy 83.5

3 hidden layers, 40, 25, and 10 neurons (1416 parameters)
Training accuracy 100.0
Validation accuracy 100.0
```

**Comment by TaehoSon-hds** [comment]
- [Adam: A Method for Stochastic Optimization](https://arxiv.org/abs/1412.6980)
- [An overview of gradient descent optimization algorithms](https://www.ruder.io/optimizing-gradient-descent/)
- [Lecture 4 | Introduction to Neural Networks](https://www.youtube.com/watch?v=d14TUNcbn1k&list=PLC1qU-LWwrF64f4QKQT-Vg5Wr4qEE1Zxk&index=5)

**Comment by TaehoSon-hds** [comment]
https://github.com/TaehoSon-hds/10-days-of-grad/blob/5cf436159758df77df8976b816a0393b56a2655a/day2/NeuralNetwork.hs#L85-L90

- [Numeric.LinearAlgebra](https://hackage.haskell.org/package/hmatrix-0.20.2/docs/Numeric-LinearAlgebra.html)
```ssh
10-days-of-grad on  day2 [!] via λ 9.2.8 via ❄️  impure (ghc-shell-for-ten-days-of-grad-0.5-0-env)
❯ ghci
GHCi, version 9.2.8: https://www.haskell.org/ghc/  :? for help
ghci> import Numeric.LinearAlgebra
ghci> a = matrix 2 [1..6]
ghci> a
(3><2)
 [ 1.0, 2.0
 , 3.0, 4.0
 , 5.0, 6.0 ]
ghci> b = matrix 2 [7, 8]
ghci> b
(1><2)
 [ 7.0, 8.0 ]
ghci> a + b
(3><2)
 [  8.0, 10.0
 , 10.0, 12.0
 , 12.0, 14.0 ]
ghci> :t a
a :: Matrix R
ghci> :t b
b :: Matrix R
ghci> :t a + b
a + b :: Matrix R
```

- [Numeric.LinearAlgebra.Static](https://hackage.haskell.org/package/hmatrix-0.20.2/docs/Numeric-LinearAlgebra-Static.html)
```ssh
10-days-of-grad on  day2 via λ 9.2.8 via ❄️  impure (ghc-shell-for-ten-days-of-grad-0.5-0-env) 
❯ ghci
GHCi, version 9.2.8: https://www.haskell.org/ghc/  :? for help
ghci> import Numeric.LinearAlgebra.Static 
ghci> :set -XDataKinds 
ghci> a = matrix [1..6] :: L 3 2
ghci> a
(matrix
 [ 1.0, 2.0
 , 3.0, 4.0
 , 5.0, 6.0 ] :: L 3 2)
ghci> b = matrix [7, 8] :: L 1 2
ghci> b
(matrix
 [ 7.0, 8.0 ] :: L 1 2)
ghci> a + b

<interactive>:7:5: error:
    • Couldn't match type ‘1’ with ‘3’
      Expected: L 3 2
        Actual: L 1 2
    • In the second argument of ‘(+)’, namely ‘b’
      In the expression: a + b
      In an equation for ‘it’: it = a + b
ghci> :t a
a :: L 3 2
ghci> :t b
b :: L 1 2
ghci> :t a + b

<interactive>:1:5: error:
    • Couldn't match type ‘1’ with ‘3’
      Expected: L 3 2
        Actual: L 1 2
    • In the second argument of ‘(+)’, namely ‘b’
      In the expression: a + b
```
